components/server/src/main/java/com/hotels/styx/server/netty/codec/NettyToStyxRequestDecoder.java

```diff
@@ -61,80 +62,108 @@
+    private Optional<FlowControllingHttpContentProducer> producer = Optional.empty();
 
     private NettyToStyxRequestDecoder(Builder builder) {
         this.uniqueIdSupplier = builder.uniqueIdSupplier;
-        this.flowControlEnabled = builder.flowControlEnabled;
         this.unwiseCharEncoder = builder.unwiseCharEncoder;
     }
 
     @Override
-    protected void decode(ChannelHandlerContext ctx, HttpObject httpObject, List<Object> out) throws Exception {
-        if (httpObject.getDecoderResult().isFailure()) {
-            throw new BadRequestException("Error while decoding request: " + httpObject, httpObject.getDecoderResult().cause());
+    protected void decode(ChannelHandlerContext ctx, HttpObject msg, List<Object> out) throws Exception {
+        if (msg.getDecoderResult().isFailure()) {
+            throw new BadRequestException("Error while decoding request: " + msg, msg.getDecoderResult().cause());
         }
 
         try {
-            if (httpObject instanceof HttpRequest) {
-                this.producer = new FlowControllingHttpContentProducer(ctx, this.flowControlEnabled);
-                Observable<ByteBuf> contentObservable = Observable.create(contentSubscriber -> {
-                    contentSubscriber.setProducer(this.producer);
-                    this.producer.subscriptionStart(contentSubscriber);
-                });
-
-                HttpRequest request = (HttpRequest) httpObject;
-                LiveHttpRequest styxRequest = toStyxRequest(request, contentObservable);
+            if (msg instanceof HttpRequest) {
+                ctx.channel().config().setAutoRead(false);
+                ctx.channel().read();
+
+                Publisher<Buffer> contentPublisher = new ContentPublisher(ctx.channel().eventLoop(), getNewContentProducer(ctx));
+
+                HttpRequest request = (HttpRequest) msg;
+                LiveHttpRequest styxRequest = toStyxRequest(request, contentPublisher);
                 out.add(styxRequest);
-            } else if (httpObject instanceof HttpContent && this.producer != null) {
-                this.producer.onNext(content(httpObject));
 
-                if (httpObject instanceof LastHttpContent) {
-                    this.producer.onCompleted();
+            }
+            if (msg instanceof HttpContent) {
+                ByteBuf content = ((ByteBufHolder) msg).content();
+                if (content.isReadable()) {
+                    getContentProducer(ctx).newChunk(retain(content));
+                }
+                if (msg instanceof LastHttpContent) {
+                    getContentProducer(ctx).lastHttpContent();
                 }
+
             }
         } catch (BadRequestException ex) {
             throw ex;
         } catch (Exception ex) {
-            throw new BadRequestException(ex.getMessage() + " in " + httpObject, ex);
+            throw new BadRequestException(ex.getMessage() + " in " + msg, ex);
         }
```

Comment:
I considered that.  I think the exception will be handled by Netty, which will pass it onto the exceptionCaught method, which calls the FSM and then propagates it to the HttpPipelineHandler.  Or have I not understood it correctly?

Reply:
Hi yes. The exception will propagate. And ideally the pipeline handler responds by closing the connection. It is still better to be defensive, and move the FSM to TERMINATED. That guarantees that any extra pending content chunk events will be correctly dealt with. Also the FSM state would be consistent with the error, in case someone ever needs to debug it.

Reply:
The exceptionCaught method makes the same call to the FSM, so it will move it to terminated.